core/src/main/scala/chisel3/package.scala

```diff
@@ -412,4 +413,25 @@ package object chisel3 {
     "this feature will not be supported as part of the migration to the MLIR-based FIRRTL Compiler (MFC). For more information about this migration, please see the Chisel ROADMAP.md."
 
   final val deprecatedPublicAPIMsg = "APIs in chisel3.internal are not intended to be public"
+
+  /** Contains universal methods for target accesses.
+    */
+  trait HasTarget {
+    def toTarget:         ReferenceTarget
+    def toAbsoluteTarget: ReferenceTarget
+    def toRelativeTarget(root: Option[BaseModule]): ReferenceTarget
+  }
+
+  object HasTarget {
+
+    /** This wrapping hides the actual object, ensuring users only have access
+      * to the target methods (instead of the type of the underlying object).
+      */
+    def wrap(t: HasTarget): HasTarget = new HasTarget {
+      def toTarget = t.toTarget
+      def toAbsoluteTarget = t.toAbsoluteTarget
+      def toRelativeTarget(root: Option[BaseModule]) = t.toRelativeTarget(root)
+    }
+
+  }
```

Comment:
```suggestion
  sealed trait HasTarget {
    def toTarget:         ReferenceTarget
    def toAbsoluteTarget: ReferenceTarget
    def toRelativeTarget(root: Option[BaseModule]): ReferenceTarget
  }

  object HasTarget {

    /** This wrapping hides the actual object, ensuring users only have access
      * to the target methods (instead of the type of the underlying object).
      */
    private[chisel3] def apply(t: NamedComponent): HasTarget = new HasTarget {
      def toTarget = t.toTarget
      def toAbsoluteTarget = t.toAbsoluteTarget
      def toRelativeTarget(root: Option[BaseModule]) = t.toRelativeTarget(root)
    }
  }
```
I know I talked you into the original implementation, but being a little more conservative here we can make sure users don't inherit this type and instead just make it a box type that only Chisel internals can create. We can always open up the API a bit later.